src/components/VioletEnroll/VioletEnroll.tsx

```diff
@@ -0,0 +1,200 @@
+import { useWeb3React } from '@web3-react/core'
+import { ButtonPrimary, VioletProtected } from 'components/Button'
+import { AutoColumn } from 'components/Column'
+import { getVioletEnrollmentCall } from 'hooks/useVioletAuthorize'
+import { MauveIcon } from 'nft/components/icons'
+import { useState } from 'react'
+import { ArrowRight } from 'react-feather'
+import { Text } from 'rebass'
+import { useToggleWalletModal } from 'state/application/hooks'
+import styled from 'styled-components/macro'
+import { ExternalLink } from 'theme'
+import { logErrorWithNewRelic } from 'utils/newRelicErrorIngestion'
+
+const Wrapper = styled.div`
+  width: 100%;
+  position: relative;
+  display: flex;
+  flex-flow: column;
+  align-items: center;
+`
+
+const Container = styled.div`
+  width: 100%;
+  padding: 32px 32px 25px 32px;
+  display: flex;
+  flex-flow: column;
+  align-items: center;
+`
+
+const Label = styled.div<{ color: string; backgroundColor: string }>`
+  width: 100%;
+  padding: 5px 5px 15px;
+  background-color: ${({ backgroundColor }) => backgroundColor};
+  border-radius: 16px;
+  color: ${({ color }) => color};
+`
+
+const TitleRow = styled.div`
+  align-items: center;
+  font-weight: 700;
+  display: inline-flex;
+`
+
+const Title = styled(Text)`
+  font-weight: 600;
+  font-size: 18px;
+  line-height: 24px;
+  margin-left: 7px;
+`
+
+const DetailsRow = styled.div`
+  margin-top: 8px;
+  font-size: 12px;
+  line-height: 16px;
+  color: ${({ theme }) => theme.textSecondary};
+`
+
+const StyledLink = styled(ExternalLink)`
+  color: ${({ theme }) => theme.textSecondary};
+  font-weight: 700;
+`
+
+const ShortColumn = styled(AutoColumn)`
+  margin-bottom: 10px;
+`
+
+const StyledButton = styled(ButtonPrimary)`
+  margin-top: 24px;
+  width: 100%;
+  font-weight: 600;
+`
+
+const StyledCloseButton = styled(StyledButton)`
+  background-color: ${({ theme }) => theme.backgroundInteractive};
+  color: ${({ theme }) => theme.textPrimary};
+
+  &:hover {
+    background-color: ${({ theme }) => theme.backgroundInteractive};
+    opacity: ${({ theme }) => theme.opacity.hover};
+    transition: opacity 250ms ease;
+  }
+  padding: 8px;
+`
+
+const RegisterButton = styled(VioletProtected(ButtonPrimary, 'backgroundContrast', '0.2rem'))`
+  padding: 10px;
+`
+
+const ConnectButton = styled(ButtonPrimary)`
+  font-size: 16px;
+  font-weight: 600;
+`
+
+export default function VioletEnroll({
+  onClose,
+  keepModalOpen,
+}: {
+  onClose: () => void
+  keepModalOpen: (open: boolean) => void
+}) {
+  const { account, chainId } = useWeb3React()
+  const [registering, setRegistering] = useState(false)
+  const [enrollmentResult, setEnrollmentResult] = useState<string>('')
+  const toggleWalletModal = useToggleWalletModal()
+
+  const onEnroll = async () => {
```

Comment:
If you have an async function defined here, shouldn't you also define the top level exported one as async?

Reply:
I don't think so? This is an async function that triggers when someone clicks the `Register with Violet` button. Correct me if I'm wrong, but this being async allows the enrolment to resolve without hanging the UI thread and doesn't require the functional react component to also be declared as async AFAIK.

I think maybe this would be the case if it was `await`ing an async function call inside another, that would have to be declared as async.

Reply:
You got me thinking about this as well now 🤔 😂 

If it gives you the correct functionality I believe you are right! 👍